oak/src/oak/server_test.cpp

```diff
@@ -0,0 +1,214 @@
+    throw std::runtime_error("popen() failed!");
+  }
+  while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
+    result += buffer.data();
+  }
+  return result;
+}
+std::string exec(const std::string& cmd) { return exec(cmd.c_str()); }
+
+std::string address;
+std::atomic<unsigned short> port;
+
+class Environment : public ::testing::Environment {
+ public:
+  ~Environment() override {}
+  void SetUp() override {
+    address = "127.0.0.1";
+    port = 1234;
+  }
+  void TearDown() override {}
+};
+testing::Environment* const foo_env = testing::AddGlobalTestEnvironment(new Environment);
+
+std::unique_ptr<oak::Server> create_server() {
+  int retry = 0;
+  do {
+    try {
+      auto port_ = port++;
+      std::cout << "Trying " << address << ":" << port_ << std::endl;
+      auto server = std::make_unique<oak::Server>(address, port_);
+      server->listen();
+      // if server is listening, return otherwise retry
+      if (server->is_listening()) {
+        return server;
+      } else {
+        retry++;
+      }
+    } catch (...) {
+      EXPECT_TRUE(false) << "Unexpected exception";
+    }
+  } while (retry < 100);
+  return nullptr;
+}
+
+/**
+ * @brief Tries to create a server
+ */
+TEST(oak_server, listen) { auto server = create_server(); }
+
+/**
+ * @brief Tests that stopping a not-listening server results in an exception
+ *
+ */
+TEST(oak_server, fail_not_listening) {
+  try {
+    oak::Server server;
+    server.stop();
+
+    FAIL() << "Expected std::runtime_error";
+  } catch (std::runtime_error) {
+    SUCCEED();
+  } catch (std::exception) {
+    FAIL() << "Expected std::runtime_error";
+  }
+}
+
+/**
+ * @brief Tries to GET a non-existing endpoint
+ */
+TEST(oak_server, get_nohandler) {
+  // Create server
+  auto server = create_server();
+  // Read a non-existing endpoint
+  auto address_ = server->address();
+  auto port_ = server->port();
+  const auto cmd =
+      boost::str(boost::format("curl -X GET %1%:%2%/test 2>/dev/null") % address_ % port_);
+  auto result = exec(cmd);
+
+  // Compare result
+  EXPECT_STREQ(result.c_str(), "{\"endpoints\":[],\"error\":\"cannot find handler\"}");
+}
+
+/**
+ * @brief Tries to GET an endpoint
+ */
+TEST(oak_server, get_handler) {
+  // Create server
+  auto server = create_server();
+  // Create registrar
+  oak::StaticRegistrar registrar(server.get(), "", nullptr);
+  // Register one pseudo-endpoint
+  std::vector<std::string> data{
+      "1",
+      "2",
+      "3",
+  };
+  registrar.register_handler("/simulators", cloe::handler::StaticJson(data));
+
+  // Read the pseudo-endpoint
+  auto address_ = server->address();
+  auto port_ = server->port();
+  const auto cmd =
+      boost::str(boost::format("curl -X GET %1%:%2%/simulators 2>/dev/null") % address_ % port_);
+  auto result = exec(cmd);
+
+  // Compare result
+  EXPECT_STREQ(result.c_str(), "[\"1\",\"2\",\"3\"]");
+}
+
+/**
+ * @brief Tries to POST to an endpoint which echos the data back
```

Comment:
Up till now, I have avoided putting these kind of doxygen comments on tests, because they're not end-user documentation. I like the comments though. I would put them as:
```
/*
 * Try to POST to the endpoint that echoes the data back.
 */
```
Note that I feel that this comment is a sentence, not a label.